src/cached_object_store.rs

```diff
@@ -762,6 +798,214 @@ impl LocalCacheEntry for FsCacheEntry {
+            Some(target) => target,
+            None => return 0,
+        };
+
+        match tokio::fs::remove_file(&target).await {
+            Err(err) => {
+                warn!("evictor: failed to remove the cache file: {}", err);
+                0
+            }
+            Ok(()) => {
+                debug!(
+                    "evictor: evicted cache file: {:?}, bytes: {}",
+                    target, target_bytes
+                );
+                target_bytes
+            }
+        }
+    }
+
+    pub async fn scan_cache_folder_bytes(
+        root_folder: &std::path::PathBuf,
+    ) -> Result<usize, Box<dyn std::error::Error>> {
+        let mut total_bytes = 0;
+        let iter = WalkDir::new(root_folder).into_iter();
+        for entry in iter {
+            let entry = entry?;
+            let metadata = tokio::fs::metadata(entry.path()).await?;
+            total_bytes += metadata.len();
+        }
+        Ok(total_bytes as usize)
+    }
+
+    // pick a file to evict, return None if no file is picked. it takes a pick-of-2 strategy, randomized pick
+    // two of files, compare their last access time, and evict the older one.
+    async fn pick_evict_target(&self) -> Option<(std::path::PathBuf, u64)> {
+        // extend the evict_buffer if it's empty. it will randomized iterate the cache folder,
+        // and randomly pick 1000 files into the evict_buffer.
+        let mut rng = rand::rngs::StdRng::from_entropy();
+        if self.evict_buffer.lock().await.len() < 2 {
+            let iter = WalkDir::new(&self.root_folder).into_iter();
+            for entry in iter.choose_multiple(&mut rng, 1000) {
+                let entry = match entry {
+                    Ok(entry) => entry,
+                    Err(err) => {
+                        warn!("evictor: failed to walk the cache folder: {}", err);
+                        break;
+                    }
+                };
+                self.evict_buffer
+                    .lock()
+                    .await
+                    .push_back(entry.path().to_path_buf());
+            }
+        }
+
+        // if the evict_buffer length still below 2, return None
+        if self.evict_buffer.lock().await.len() < 2 {
+            return None;
+        }
+
+        // pop two files from the evict_buffer
+        let (path0, path1) = {
+            let mut guard = self.evict_buffer.lock().await;
+            match (guard.pop_front(), guard.pop_front()) {
+                (Some(path0), Some(path1)) => (path0, path1),
+                _ => return None,
+            }
+        };
+
+        // read the accessed time & length of the two files
+        let (metadata0, metadata1) = match (
+            tokio::fs::metadata(&path0).await,
+            tokio::fs::metadata(&path1).await,
+        ) {
+            (Ok(metadata0), Ok(metadata1)) => (metadata0, metadata1),
+            (Err(err), _) | (_, Err(err)) => {
```

Comment:
do we care about the case when just 1 of these files produced a metadata object? With this current implementation, it is a 2 files or bust approach which  can be wasteful (considering that we had to do a bit of work to get to this point in the execution span of this function... and the goal is to evict just one file anyway). Instead, it should be tolerable if just 1 of them return a metadata object we can work with and just move on to evicting with it without having to decide which of 2 files need evicting.

Reply:
imho it would be a extremely rare to get error on reading the metadata of the files (maybe on the cases like a race condition where some others deleting the files in the same time?)

i think we could assume these errors almostly never happens in the real world, maybe we can keep the error handling logic simpler here?